# Проект по SQL

### Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. 

В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. 

Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


### Описание данных

Таблица books

Содержит данные о книгах:

- book_id — идентификатор книги;

- author_id — идентификатор автора;

- title — название книги;

- num_pages — количество страниц;

- publication_date — дата публикации книги;

- publisher_id — идентификатор издателя.


Таблица authors

Содержит данные об авторах:

- author_id — идентификатор автора;

- author — имя автора.


Таблица publishers

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;

- publisher — название издательства;


Таблица ratings

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;

- book_id — идентификатор книги;

- username — имя пользователя, оставившего оценку;

- rating — оценка книги.


Таблица reviews

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;

- book_id — идентификатор книги;

- username — имя автора обзора;

- text — текст обзора.


### Задания
1. Посчитайте, сколько книг вышло после 1 января 2000 года;

2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


### Цели исследования:

1. проанализировать базу данных.

2. Сформулировать ценностное предложение для нового продукта.

**Импортируем нужные библиотеки**

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

**Подключимся к базе данных**

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**Исследуем таблицы и выведем первые строки каждой таблици**

In [3]:
# Таблица books
query = '''SELECT * 
           FROM books 
           LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# Таблица authors
query = '''SELECT * 
           FROM authors 
           LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
# Таблица publishers
query = '''SELECT * 
           FROM publishers 
           LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
# Таблица ratings
query = '''SELECT * 
           FROM ratings 
           LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
# Таблица reviews
query = '''SELECT * 
           FROM reviews 
           LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

1. **Посчитаем, сколько книг вышло после 1 января 2000 года**

In [8]:
query = '''
           SELECT COUNT(*) 
           FROM books 
           WHERE publication_date > '2000-01-01';
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

- После 1 января 2000 года вышло 819 книг

2. **Для каждой книги посчитаем количество обзоров и среднюю оценку**

In [9]:
query = '''
           SELECT books.title, 
           COUNT(DISTINCT reviews.review_id) AS count_reviews, 
           AVG(ratings.rating) AS avg_rating
           FROM books
           LEFT JOIN reviews ON books.book_id = reviews.book_id
           LEFT JOIN ratings ON books.book_id = ratings.book_id
           GROUP BY books.book_id;
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  count_reviews  \
0                                         'Salem's Lot              2   
1                   1 000 Places to See Before You Die              1   
2    13 Little Blue Envelopes (Little Blue Envelope...              3   
3    1491: New Revelations of the Americas Before C...              2   
4                                                 1776              4   
..                                                 ...            ...   
995           Wyrd Sisters (Discworld  #6; Witches #2)              3   
996                        Xenocide (Ender's Saga  #3)              3   
997                                    Year of Wonders              4   
998                        You Suck (A Love Story  #2)              2   
999  Zen and the Art of Motorcycle Maintenance: An ...              4   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 3 columns]

- count_reviews - количество обзоров по каждой книге.
- avg_rating - средняя оценка по каждой книге.

3. **Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так исключим из анализа брошюры**

In [10]:
query = '''
           SELECT publishers.publisher, 
           COUNT(books.book_id) AS num_books
           FROM books
           JOIN publishers ON books.publisher_id = publishers.publisher_id
           WHERE books.num_pages > 50
           GROUP BY publishers.publisher_id
           ORDER BY num_books DESC
           LIMIT 1;
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  num_books
0  Penguin Books         42

- Penguin Books - издательство, которое выпустило наибольшее число книг толще 50 страниц. 
- Penguin Books Выпустило 42 книги.

4. **Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками**

In [11]:
query = '''
           SELECT authors.author, AVG(ratings.rating) AS avg_rating
           FROM authors   
           JOIN books ON authors.author_id = books.author_id   
           JOIN ratings ON books.book_id = ratings.book_id   
           WHERE books.book_id IN   
           (SELECT book_id
           FROM ratings
           GROUP BY book_id
           HAVING COUNT(*) >= 50)
           GROUP BY authors.author_id
           ORDER BY avg_rating DESC
           LIMIT 1;
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

- J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой книг.

5. **Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [12]:
query = '''
           SELECT AVG(subquery.review_count) AS avg_review_count
FROM (
  SELECT COUNT(*) AS review_count
  FROM reviews
  WHERE username IN (
    SELECT username
    FROM (
      SELECT username, COUNT(*) AS rating_count
      FROM ratings
      GROUP BY username
      HAVING COUNT(*) > 48
    ) AS rating_subquery
  )
  GROUP BY username
) AS subquery;

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg_review_count
0              24.0

- Пользователи, которые поставили больше 48 оценок в среднем сделали 24 обзора.

### Вывод

- После 1 января 2000 года вышло 819 книг


- Penguin Books - издательство, которое выпустило наибольшее число книг толще 50 страниц. Оно Выпустило 42 книги.


- J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой книг.


- Пользователи, которые поставили больше 48 оценок в среднем сделали 24 обзора.

### Рекомендации

- Выпускать большее количество новых книг разных жанров.


- Рекламировать сервис на разных площадках. 


- Поощрять клиентов за обзоры на книги. Например делать скидку на подписку.